### PR TITLE
Plugins: 'channel_opened' notification type

### PR DESCRIPTION
**Pull request by darosior**
This adds a new notification type : `channelopened`.

This notification is emitted if a peer successfully opened a channel to us, and sends the id of the peer, the funding amount, and the feerate.

### CODE REVIEW

File: tests/plugins/misc_notifications.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env python3
+"""Plugin to be used to test miscellaneous notifications.
+
+Only used for 'channel_opened' for now.
```
Comment (ZmnSCPxj) [inline tests/plugins/misc_notifications.py]: `channelopened` is what is actually used.

File: tests/test_plugin.py [hunk 2]
```diff
@@ -476,3 +476,15 @@ def test_warning_notification(node_factory):
...
+def test_channelopened_notification(node_factory):
+    """
+    Test the 'channelopened' notification sent at channel funding success.
+    """
+    opts = [{}, {'plugin': 'tests/plugins/misc_notifications.py'}]
+    amount = 10**6
+    l1, l2 = node_factory.line_graph(2, fundchannel=True, fundamount=amount,
+                                     opts=opts)
+    l2.daemon.wait_for_log(r"A channel to us was opened by {}, with an amount"
+                           " of {}msat*".format(l1.info["id"], amount*1000))
```
Comment (ZmnSCPxj) [inline tests/test_plugin.py]: Triggers a warning:

>      tests/test_plugin.py:490:70: E226 missing whitespace around arithmetic operator

**Comment by darosior** [comment]
Thanks for the review @ZmnSCPxj, I corrected and will resolve conflict in the upcoming rebase.

File: doc/PLUGINS.md [hunk 3]
```diff
@@ -200,6 +200,20 @@ corresponding payloads are listed below.
 
 ### Notification Types
 
+#### `channelopened`
+
+A notification for topic `channelopened` is sent if a peer successfully funded a channel
```
Comment (cdecker) [inline doc/PLUGINS.md]: We seem to get inconsistent with the naming of hooks:

 - `htlc_accepted`
 - `invoice_payment`
 - `db_write`
 - `channelopened`

I'd suggest adding the `_` for consistency's sake :-)

Comment (darosior) [reply doc/PLUGINS.md]: Ok :-)

**Comment by darosior** [comment]
Corrected the naming and the include duplication.

File: tests/test_plugin.py [hunk 4]
```diff
@@ -494,3 +494,15 @@ def test_invoice_payment_notification(node_factory):
     l2.daemon.wait_for_log(r"Received invoice_payment event for label {},"
                            " preimage {}, and amount of {}msat"
                            .format(label, preimage, msats))
+
+
+def test_channel_opened_notification(node_factory):
+    """
+    Test the 'channelopened' notification sent at channel funding success.
```
Comment (cdecker) [inline tests/test_plugin.py]:
```suggestion
    Test the 'channel_opened' notification sent at channel funding success.
```

File: tests/plugins/misc_notifications.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+"""
+
+from lightning import Plugin
+
+plugin = Plugin()
+
+
+@plugin.subscribe("channel_opened")
+def on_channelopened(plugin, channel_opened):
```
Comment (cdecker) [inline tests/plugins/misc_notifications.py]:
```suggestion
def on_channel_opened(plugin, channel_opened):
```

File: lightningd/notification.c [hunk 6]
```diff
@@ -76,3 +77,17 @@ void notify_invoice_payment(struct lightningd *ld, struct amount_msat amount,
 	jsonrpc_notification_end(n);
 	plugins_notify(ld->plugins, take(n));
 }
+
+void notify_channel_opened(struct lightningd *ld, struct node_id *node_id,
+			 struct amount_sat *funding_sat, u32 *feerate_per_kw)
+{
+	struct jsonrpc_notification *n =
+	    jsonrpc_notification_start(NULL, notification_topics[4]);
```
Comment (cdecker) [inline lightningd/notification.c]: The use of indices here can lead to unwanted renamings if the `notification_topics` array get reordered. Prefer adding an inline string here.

**Comment by darosior** [comment]
Arf, forgot these ones

File: doc/PLUGINS.md [hunk 7]
```diff
@@ -200,6 +200,20 @@ corresponding payloads are listed below.
...
+#### `channel_opened`
+
+A notification for topic `channel_opened` is sent if a peer successfully funded a channel
+with us. It contains the peer id, the funding amount, and the feerate (per kw).
+```
+{
+	"channel_opened": {
+		"id": "03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f",
+		"funding_satoshis": "100000000msat",
+		"feerate": 7500,
```
Comment (cdecker) [inline doc/PLUGINS.md]: `feerate` seems like an odd detail to add here, but something like the funding txid would be useful here imho.

**Comment by darosior** [comment]
Done :)

File: lightningd/notification.c (hunk 6 above)
Comment (ZmnSCPxj) [reply lightningd/notification.c]: Probably appropriate to modify all of the notifications in a separate commit, then, as there is existing code which uses the same pattern.

File: doc/PLUGINS.md (hunk 7 above)
Comment (darosior) [reply doc/PLUGINS.md]: Ok, I'll replace the feerate by the txid

**Comment by darosior** [comment]
> Actual commit messages use `channelopened` and `openchannel`, should use `channel_opened`?

Yes, I had not the time to change it again yesterday.

While including the `funding_txid` as requested by @cdecker I moved the notification to the `opening_fundee_finished()` function, which (I think) makes more sense. I also added a drive-by commit to remove the use of `notification_topics` indices in notification functions. The notification now sends :
```
{
	"channel_opened": {
		"id": "03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f",
		"funding_satoshis": "100000000msat",
		"funding_txid": "4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b",
		"funding_locked": false
	}
}
```

File: lightningd/opening_control.c [hunk 8]
```diff
@@ -550,7 +550,7 @@ static void opening_funder_finished(struct subd *openingd, const u8 *resp,
 	/* Needed for the success statement */
 	derive_channel_id(&fc->cid, &channel->funding_txid, funding_txout);
 
-	/* Was this an external wallet initiation ? */
+		/* Was this an external wallet initiation ? */
```
Comment (ZmnSCPxj) [inline lightningd/opening_control.c]: Unnecessary change?

**Comment by darosior** [comment]
Corrected, thanks.

**Comment by cdecker** [comment]
ACK https://github.com/ElementsProject/lightning/pull/2840/commits/1221981e36b7357aa78e81f7bb9eb3a03bca9b9a
